docs/docs/usage/host-functions.md

```diff
@@ -114,3 +114,79 @@ logIt.apply();
 // should print "Hello, World!" 10 times
 ```
 
+## Host Function Modules
+
+Instead of writing host functions by hand, you can write a class containing annotated methods
+and let the Chicory annotation processor generate the host functions for you. This is especially
+useful when you have many host functions.
+
+```java
+@HostModule("demo")
+public final class Demo {
+    @WasmExport
+    public long add(int a, int b) {
+        return a + b;
+    }
+
+    @WasmExport // the Wasm name is random_get
+    public void randomGet(Memory memory, int ptr, int len) {
+        byte[] data = new byte[len];
+        random.nextBytes(data);
+        memory.write(ptr, data);
+    }
+
+    public HostFunction[] toHostFunctions() {
+        return Demo_ModuleFactory.toHostFunctions(this);
+    }
+}
+```
+
+The `@HostModule` annotation marks the class as a host module and specifies the module name for
+all the host functions. The `@WasmExport` annotation marks a method as host function and optionally
+specifies the name of the function. If the name is not specified, then the Java method name is
+converted from camel case to snake case, as is a common convention in Wasm.
+
+The `Demo_ModuleFactory` class in `toHostFunctions()` is generated by the annotation processor.
+
+Host functions must be instance methods of the class. Static methods are not supported.
+This is because host functions will typically interact with instance state in the host class.
+
+To use the host module, you need to instantiate the host module and fetch the host functions:
+
+```java
+var demo = new Demo();
+var imports = new HostImports(demo.toHostFunctions());
+```
+
+### Type conversions
+
+The following conversions are supported:
+
+| Java Type         | Wasm Type  |
+|-------------------|------------|
+| `int`             | `i32`      |
+| `long`            | `i64`      |
+| `float`           | `f32`      |
+| `double`          | `f64`      |
+
+### Enabling the Annotation Processor
+
+In order to use host modules, you need to configure the Java compiler to include the Chicory
+`function-processor` as an annotation processor. Exactly how this is done depends on the build
+system you are using. This is how to do it with Maven:
+
+```xml
+<plugin>
+  <groupId>org.apache.maven.plugins</groupId>
+  <artifactId>maven-compiler-plugin</artifactId>
+  <configuration>
+    <annotationProcessorPaths>
+      <path>
+        <groupId>com.dylibso.chicory</groupId>
+        <artifactId>function-processor</artifactId>
+        <version>DOCS_PLACEHOLDER_VERSION</version>
```

Comment:
We should be able to remove this version and instead, document the usage of the `BOM`

Reply:
This can be done as a follow up, @kanatohodets let me know if you are interested in documenting this "super Java specific" aspect of the project, [ref](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#bill-of-materials-bom-poms).

This is an example usage:
- you include the BOM in `dependencyManagement`: https://github.com/andreaTP/opa-chicory/blob/9931be414b6d5f9faca25fa07445b868ab443b25/pom.xml#L66-L76
- you declare all artifact dependencies without the version: https://github.com/andreaTP/opa-chicory/blob/9931be414b6d5f9faca25fa07445b868ab443b25/core/pom.xml#L12-L15